Sight_++_Solution/Intel_Sight_PP/src/object_tracking.cpp

```diff
@@ -1,118 +1,103 @@
-#include <iostream>
-#include <string>
-#include <vector>
-#include <opencv2/tracking.hpp>
-#include "classification_result.hpp"
+#include "object_tracking.hpp"
 
-struct TrackingItem
-{
-	std::string name;
-	std::vector<double> distance;
-	std::vector<cv::Rect2d> rec;
-	cv::Ptr<cv::Tracker> tracker;
-	int counter;
-	int track_point;
-	bool lock;
-	double speed;
-	int last_seen;
-
-	TrackingItem(const std::string name, const std::vector<double> distance, const std::vector<cv::Rect2d> rec, const int counter, const int track_point, const bool lock, const double speed,const int last_seen, const cv::Mat color_matrix, const cv::Rect2d init_rec) :
-		name(name), distance(distance), rec(rec),  counter(counter), track_point(track_point), lock(lock), speed(speed), last_seen(last_seen)
-	{
-		tracker = cv::TrackerMOSSE::create();
-		tracker->init(color_matrix, init_rec);
+void ObjectTracking::frame_update(cv::Mat color_matrix) {
+	update_items.clear();
+	for (int i = 0; i < items.size(); i++) {
+		cv::Rect2d update_rec;
+		items[i].tracker->update(color_matrix, update_rec);
+		update_items.push_back(update_rec);
 	}
+}
 
-};
-struct ObjectTracking {
-	std::vector<TrackingItem> items;
-	const int track_frame = 5;
-	std::vector<cv::Rect2d> update_items;
-	std::vector<double> frame_time;
-	std::vector<ClassificationItem> result_objects;
-
-	void frame_update(cv::Mat color_matrix) {
-		update_items.clear();
-		for (int i = 0; i < items.size(); i++) {
-			cv::Rect2d update_rec;
-			items[i].tracker->update(color_matrix,update_rec);
-			update_items.push_back(update_rec);
+std::vector<ClassificationItem> ObjectTracking::post_process(double stamp) {
+	result_objects.clear();
+	for (int i = 0; i < items.size(); i++) {
+		if (!items[i].lock) {
+			items[i].track_point--;
+			if (items[i].track_point <= 0) {
+				items.erase(items.begin() + i);
+			}
+		}
+		else {
+			items[i].lock = false;
 		}
 	}
-	std::vector<ClassificationItem> post_process(double stamp) {
-		result_objects.clear();
-		for (int i = 0; i < items.size(); i++) {
-			if (!items[i].lock) {
-				items[i].track_point--;
-				if (items[i].track_point <= 0) {
-					items.erase(items.begin() + i);
-				}
+	frame_time.push_back(stamp);
+	if (frame_time.size() > track_frame + 1) {
+		frame_time.erase(frame_time.begin());
+	}
+	for (int i = 0; i < items.size(); i++) {
```

Comment:
Can't you use a Map to store the items? 
Where the item id is the key, and the value is the item/pointer to it?

That may require a not insignificant rewrite though 😄

Reply:
Ye that would work too. Though it would make it difficult of id tracking and reuse. Unless you just monitor empty slots i guess. That would work well too I think. Though the iteration through all tracked items at all times may make the changes not that significant?